Switch to terser-webpack-plugin

edmorley opened the pull request:
Since the webpack 4 default of `uglifyjs-webpack-plugin` uses `uglify-es`, which has been deprecated in favour of `terser`. webpack 5 is due to use `terser-webpack-plugin`, at which point we can remove this.

`terser` includes multiple correctness and performance fixes not present in `uglify-es`.

Fixes #1146.

edmorley commented on packages/web/README.md:
```diff
@@ -536,38 +536,31 @@ module.exports = {
+(which replaces `uglifyjs-webpack-plugin`). To customise the options passed to `TerserPlugin`
 or even use a different minifier, override `optimization.minimizer`.
 
-Note: If switching to [babel-minify-webpack-plugin](https://github.com/webpack-contrib/babel-minify-webpack-plugin)
```
I've intentionally removed this reference to babel-minify-webpack-plugin, since it's still virtually unmaintained/buggy/slow. If that changes in the future we can re-add.

edmorley commented:
I was thinking about this PR some more today. Since we manually configure `terser-webpack-plugin`, we have to choose whether to set `sourceMaps` to true/false. This is fine if people use the in-built web preset `devtool` option, however for Neutrino 8 upgraders it's likely they may have:

```js
module.exports = {
  use: [
    ['@neutrinojs/react', {
      // ...
    }],
    (neutrino) => {
      if (process.env.NODE_ENV === 'production') {
        neutrino.config.devtool('source-map');
      }
    },
  ]
};
```

In this case, we'll pass `sourceMaps: false` to `terser-webpack-plugin`.

The only thing I think we can do, is to perhaps add a warning (or throw) in the webpack() handler, like we do for the `vendor` entrypoint check, which checks whether `devtool` and `sourceMap` are out of sync (if the `terser` plugin found).

@eliperelman, thoughts?

eliperelman commented:
@edmorley I think that solution will work, but we should probably use a warning since it may be desired by the user to have them out of sync for some reason.

edmorley commented:
I've updated the PR:
* rebased on `master`'s webpack-chain v5 upgrade
* added a check for misconfiguration in `handlers.js` (see comment there for why throwing rather than a `console.error()`)
* added migration guide changes
